### PR TITLE
signature change for `node_link` functions: for issue #5787

### LINKED ISSUE #5787: `node_link` functions `attrs` kwarg
`node_link_graph` and `node_link_data` have a similar API to the `cytoscape` functions, where there is a `attrs` kwarg which expects a dictionary containing specific keys that is then unpacked internally.

This API is a bit un-Pythonic and was removed in favor of keyword arguments - see the related discussion in #4199. It's worth checking if the `node_link` functions wouldn't benefit from a similar treatment.

### PR DESCRIPTION
**Pull request by brocla**
Signature change for `node_link` functions resolves #5787.
Attributes are passed by keyword instead of a dictionary.

Not yet complete.
*** The deprecation warnings need to be edited to include the target version for the change ***

Tests
- Tests were added to check the new signature with keywords.
- A test was added to verify that the warning message is triggered when the old style signature is used.
- Tests that check the old style signatures were kept, but were tagged with TODO comments for deletion.

Docs
- The doc strings were updated with deprecation notices
- *** The examples in the doc strings still need updating.

### CODE REVIEW

**Review by dschult — APPROVED** [review]
Nicely done! 
This is a very nice rewrite to allow the keyword arguments in place of a dict argument with keys for each variable.

@rossbar @jarrodmillman  If we put this in a v2.8.6 release can the deprecation be removed in v3.0 or do we have to wait for v3.1?

My approval assumes a choice for version number when we remove the old code/tests/warnings. I hope it can be v3.0. :}

**Comment by brocla** [comment]
Note: Even thought this PR passed its tests, it cannot be merged yet. 
The version number, when the deprecated code will be removed, still needs to be added to the docstrings.

File: networkx/readwrite/json_graph/node_link.py
Comment (MridulS) [inline networkx/readwrite/json_graph/node_link.py:121]: Maybe it's an overkill but should there be a check here if there is a mismatch between the `attrs` attribute dict and the keyword args? Something like could be a bit confusing 
``` python
>>> data1 = nx.node_link_data(G, attrs={'source':'s'}, source='t')
>>> data1
{'directed': False,
 'multigraph': False,
 'graph': {},
 'nodes': [{'name': 'A'}, {'name': 'B'}],
 'links': [{'s': 'A', 'target': 'B'}]}

```

This is probably an overkill as `attrs` would be visibly deprecated if someone tries to use the other keyword arguments

File: networkx/readwrite/json_graph/node_link.py
Comment (MridulS) [inline networkx/readwrite/json_graph/node_link.py:53]: FWIW, I think it should be deprecated in 2.8.6 and removed in 3.1 (we should wait for atleast 2 releases before removing stuff)

File: networkx/readwrite/json_graph/node_link.py
Comment (brocla) [reply networkx/readwrite/json_graph/node_link.py:121]: What if I documented the behavior by adding this line to the docstring...

> Until the `attrs=` keyword is removed, it takes precedence over the new keywords, should they both be used in the same call (not recommended).

Comment (MridulS) [reply networkx/readwrite/json_graph/node_link.py:125]: Sounds good to me :)

**Review by MridulS — APPROVED** [review]
Thanks @brocla!
All looks good to me :)
Before merging I'll wait for someone else to +1 the NX versions for deprecation and removal :)
@rossbar @dschult ?

**Comment by dschult** [comment]
The [deprecation policy](https://networkx.org/documentation/stable/developer/deprecations.html) (I found it! yay) says 
> Usually, our policy is to put in place a deprecation cycle over two minor releases (e.g., if a deprecation warning appears in 2.3, then the functionality should be removed in 2.5). For major releases we usually require that all deprecations have at least a 1-release deprecation cycle (e.g., if 3.0 occurs after 2.5, then all removed functionality in 3.0 should be deprecated in 2.5).

>Note that these 1- and 2-release deprecation cycles for major and minor releases is not a strict rule and in some cases, the developers can agree on a different procedure upon justification (like when we can’t detect the change, or it involves moving or deleting an entire function for example).

In this case, we would deprecate this in 2.8.6...  so it is not deprecated in 2.8.  So I think it really should be removed in v3.1.  We could argue for a special case, but this is easier and less dramatic for users.

**Comment by brocla** [comment]
Great. Thanks for the reviews.
 I agree with deprecating in 2.8.6 and removing at 3.1.   I have put those numbers in the docstrings and warning messages.

Also, I found some instructions on deprecating in the [Contributor Guide.](https://networkx.org/documentation/stable/developer/contribute.html)  I added a deprecation warning to `networkx/conftest.py` and set a reminder in `doc/developer/deprecations.rst`

I did not put a note in `doc/release/release_dev.rst` because that file is for 3.0, and this change is for 3.1

**Comment by rossbar** [comment]
+1 for the changes - just wanted to comment here that I'd actually advocate for removal in 3.2 rather than 3.1. *Technically* deprecations shouldn't be introduced in patch releases, so the next non-patch release that would have the deprecation warning would be 3.0 which, according to the policy, would mean removing in 3.2.

It's a very minor point and I don't feel too strongly about it, but just thought I'd bring it up. If others agree I'm happy to make the change, but if others are happy with how it is that's fine with me too!

**Comment by dschult** [comment]
Sorry @rossbar for pulling the trigger before you had a chance to look at this!  

I'm fine with shifting the deprecation date to the 3.2 release. The rules for deprecation are not clear to me, especially once we started making the patch releases more substantial, (which I suspect is being done to make sure 2.8 has all the recent bug fixes so that people who don't want to shift to 3.0 can stick with 2.8).

But, just to make sure I understand correctly: the stuff we deprecated before v2.8 could be removed in v3.0 because that is a major release. The stuff we deprecate now -- before v3.0 and patched into 2.8.6 -- would be removed in v3.2 because we need 2 minor releases or one major release between announcement and removal.

**Comment by rossbar** [comment]
> But, just to make sure I understand correctly: the stuff we deprecated before v2.8 could be removed in v3.0 because that is a major release. The stuff we deprecate now -- before v3.0 and patched into 2.8.6 -- would be removed in v3.2 because we need 2 minor releases or one major release between announcement and removal.

This is my understanding, yes. Under normal circumstances I don't think we'd want to be introducing deprecations in patch releases, but as you noted the patches for 2.8 have been *special* in some sense since we're doing a lot more backporting than ususal. Given all this *I think* the above interpretation makes sense, but as I say, I don't feel particularly strongly about it!

**Pull request merged**
